### PR TITLE
malloc: reimplement allocator

### PR DESCRIPTION
**Pull request by reusee**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #15479 

## What this PR does / why we need it:

malloc: add Allocator interface

malloc: add Allocator tests

malloc: more refinements

malloc: fix testAllocator

malloc: add ShardedAllocator

malloc: remove old malloc codes

memorycache: fix allocation

### CODE REVIEW

File: pkg/common/malloc/class_allocator.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+		select {
+
+		case ptr := <-f.buffer2:
+			// from buffer2
+			clear(unsafe.Slice((*byte)(ptr), f.size))
+			return ptr, f
+
+		default:
+			// allocate new
+			data, err := unix.Mmap(
```
Comment (fengttt) [inline pkg/common/malloc/class_allocator.go:112]: why we use Mmap?

File: pkg/common/malloc/class_allocator.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+func (f *fixedSizeMmapAllocator) Deallocate(ptr unsafe.Pointer) {
+	select {
+
+	case f.buffer1 <- ptr:
+		// buffer in buffer1
+
+	default:
+
+		// madvise
+		if err := unix.Madvise(
```
Comment (fengttt) [inline pkg/common/malloc/class_allocator.go:141]: Here again, I am not sure you want to deal with Advise

At least we need to have evidence that this is a good idea.

File: pkg/fileservice/memorycache/cache.go [hunk 3]
```diff
@@ -60,11 +61,17 @@ func NewCache(
 		}
 	}
 	c.l = lrucache.New(capacity, setFunc, getFunc, evictFunc)
+	c.allocator = malloc.NewShardedAllocator(
```
Comment (fengttt) [inline pkg/fileservice/memorycache/cache.go:64]: Shouldn't here be a config that we decide which allocator to use?

**Review by fengttt — CHANGES_REQUESTED** [review]
Why unix.Mmap is used?    Why not just use go new/make?   Do we have reason to believe mmap is any better?

Next, I think we need the config to switch allocators.   For experiment etc.    Otherwise what is the point of having more than one allocator?

Then, we are missing the check that all Allocated memory are Freed.   

Finally, please add another Allocator.    Just call C.malloc and C.free.   By now I am very curious of all your work, if benchmarked with C.malloc/free, how much better and worse under different workload.

File: pkg/common/malloc/class_allocator.go (hunk 1 above)
Comment (reusee) [reply pkg/common/malloc/class_allocator.go:112]: 下面评论回复

File: pkg/common/malloc/class_allocator.go (hunk 2 above)
Comment (reusee) [reply pkg/common/malloc/class_allocator.go]: 下面评论回复

File: pkg/fileservice/memorycache/cache.go (hunk 3 above)
Comment (reusee) [reply pkg/fileservice/memorycache/cache.go:64]: 没必要暴露内部实现到配置文件。

**Comment by reusee** [comment]
> Why unix.Mmap is used? Why not just use go new/make? Do we have reason to believe mmap is any better?
> 
> Next, I think we need the config to switch allocators. For experiment etc. Otherwise what is the point of having more than one allocator?
> 
> Then, we are missing the check that all Allocated memory are Freed.
> 
> Finally, please add another Allocator. Just call C.malloc and C.free. By now I am very curious of all your work, if benchmarked with C.malloc/free, how much better and worse under different workload.

这个分配器不会带来任何 GC 开销，这是 go 的 make 不具备的。
go 的 make 分配的内存，没法立即释放，这会导致峰值内存很高，从而 oom。
madvise MADV_DONTNEED 可以立即释放内存，RSS马上会降低，这是使用 mmap + madvise 的理由。
go的分配器也是这样做的。
而且这个最大支持8GB的分配，之前的方案实现不了。

这个设计受这篇论文启发，请看 2.1 节：https://www.semanticscholar.org/paper/Umbra%3A-A-Disk-Based-System-with-In-Memory-Neumann-Freitag/88a0ce733f712e742733362623aa0a27710a5b9f 

这个可以解决OOM问题。

**Review by fengttt — APPROVED** [review]

**Pull request merged**